### PR TITLE
Fix mse_loss/smooth_l1_loss allocating oversized storage for scalar output

### LINKED ISSUE #185647: Unexpected storage from MSE loss output
### 🐛 Describe the bug

MSE loss output takes up more storage than expected even though its not a view.

```python
import torch
import torch.nn.functional as F

x = torch.rand(3, 256, 256)
recon = torch.rand(3, 256, 256)

l2_loss = F.mse_loss(x, recon, reduction='mean')
print(l2_loss)
print(l2_loss.untyped_storage().nbytes())
print(l2_loss._base)
```
Output:
```
tensor(0.1674)
786432
None
```

This behavior doesn't occur with other reductions and some other losses (e.g., F.l1_loss).

### Versions

PyTorch version: 2.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.13 (main, Jun  3 2025, 18:38:25) [Clang 17.0.0 (clang-1700.0.13.3)] (64-bit runtime)
Python platform: macOS-15.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M3

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.5
[pip3] torch==2.9.0
[pip3] torch-fidelity==0.3.0
[pip3] torchaudio==2.9.0
[pip3] torchcodec==0.9.1
[pip3] torchvision==0.24.0
[pip3] torchviz==0.0.3
[conda] Could not collect

### PR DESCRIPTION
**Pull request by gaurav-redhat**
## Summary
- Fix F.mse_loss() and F.smooth_l1_loss() with reduction='mean' or 'sum' allocating storage equal to the full input tensor size instead of scalar size (4 bytes for float32).
- Root cause: the meta function called build_borrowing_binary_op() (full-size alloc) then resize_({}) (only changes shape, not storage). Fix uses set_output_raw_strided() to directly allocate scalar-sized output.
- Added regression test test_loss_reduction_storage_size.

Fixes #185647

## Test plan
- Added test_loss_reduction_storage_size in test/test_nn.py
- Verified correctness for mean/sum/none reductions
- Verified mixed-dtype (float32 + float64) type promotion
- Verified backward pass is unaffected

### CODE REVIEW

**Review by zou3519 — CHANGES_REQUESTED** [review]
test failures look real
